### ISSUE TITLE
dataset split: /var/spool/sge/schoolshark/job_scripts/6843341: line 13: 51804 Illegal instruction     (core dumped)

### ISSUE THREAD

**Issue by bastian-wur** [aldro61/kover#42]
Hi everyone,

I'm very interested in using Kover, since I regularly get bombarded with medium size datasets, where people are looking for some difference.
Therefore I'm trying to run it now on a set of WGS data, for which we already identified the cause of the difference in phenotype.
The command dataset split crashes though, and I do not know why.

My commands are the following:
`
/exports/mm-hpc/bacteriologie/bastian/tools/kover/bin/kover dataset create from-reads --genomic-data /exports/mm-hpc/bacteriologie/bastian/data/AB_resistance_new/read_overview.csv --phenotype-metadata /exports/mm-hpc/bacteriologie/bastian/data/AB_resistance_new/metadata_overview.csv --output /exports/mm-hpc/bacteriologie/bastian/data/AB_resistance_new/kover_AB_test --n-cpu 8 -x --phenotype-name "AB resistance"
/exports/mm-hpc/bacteriologie/bastian/tools/kover/bin/kover dataset split --dataset /exports/mm-hpc/bacteriologie/bastian/data/AB_resistance_new/kover_AB_test --id AB_split --train-size 0.666 --folds 2
/exports/mm-hpc/bacteriologie/bastian/tools/kover/bin/kover learn --dataset /exports/mm-hpc/bacteriologie/bastian/data/AB_resistance_new/kover_AB_test --split AB_split --model-type conjunction disjunction --p 0.1 1.0 10.0 --max-rules 5 --hp-choice cv --n-cpu 8 --progress --random-seed 72 --output-dir /exports/mm-hpc/bacteriologie/bastian/data/AB_resistance_new/kover_AB_result`

The creation of the dataset works, but the splitting doesn't, I get the following error:
`/var/spool/sge/schoolshark/job_scripts/6843341: line 13: 51804 Illegal instruction     (core dumped) /exports/mm-hpc/bacteriologie/bastian/tools/kover/bin/kover dataset split --dataset /exports/mm-hpc/bacteriologie/bastian/data/AB_resistance_new/kover_AB_test --id AB_split --train-size 0.666 --folds 2`

This seems to be somewhere in error in one of the tools the data is fed to, at least as far as I can see.
If I could get a pointer which tool/library that could be, that would be very helpful.

Thank you very much :).

Bastian

**Comment by aldro61** [comment]
Hi Bastian, 

Thanks for your interest in kover. I have never encountered this error before. The splitting code uses basic numpy and h5py functionality. Can you make sure that your versions of numpy and scipy are up to date?

`pip install --upgrade numpy h5py`

Also, are you able to run this example without errors? http://aldro61.github.io/kover/doc_example.html

Thanks!
Alex

**Comment by bastian-wur** [comment]
Hi Alex,

thanks for coming back to me so quickly.
No, I cannot run the example either, splitting also crashes.
I don't get a stack trace, but a core dump. No clue how to analyze that one.

Before running the last test, I also upgraded h5py and numpy. They're both installed via conda. Numpy is now on 1.13.3 and h5py on 2.7.1.
Which versions are on your system?

I'm trying to run kover on a cluster, using SGE for job distribution, but that doesn't have an effect, I also tried to directly execute it on one of the nodes.
The system runs with Ubuntu 14.04.5 LTS, and my python and everything related is in a conda environment.

Maybe it's really the versions, that would be the easiest to handle.

Regards,
Bastian

**Comment by aldro61** [comment]
Hi Bastian,

I run kover on a variety of computers on Linux and Mac. On my current computer (Mac), I have numpy 1.13.3 and h5py 2.7.0, also installed via Anaconda.

I suspect that it might have something to do with the popcount CPU instructions that kover uses. Are you able to run the following code?

```
from kover.learning.set_covering_machine.popcount import inplace_popcount_64
a = np.random.randint(0, 2**63 - 1, (5, 10)).astype(np.uint64)
inplace_popcount_64(a, np.ones(5, dtype=np.uint64))
print(a.sum(axis=0))
```

Thanks!
Alex

**Comment by aldro61** [comment]
Other important question: did you compile with GCC or another compiler? This is essential, since kover uses GCC builtin functions: https://github.com/aldro61/kover/blob/master/core/kover/learning/set_covering_machine/popcount.pyx#L27.

**Comment by bastian-wur** [comment]
Hi Alex,

regarding compiling: Uh, default? The log during compiling says GCC 4.8.4. I checked, that one seems outdated. I might try to upgrade it.

Regarding your test script: Yes, that works, the output is
[2 3 3 4 4 4 2 2 2 4]

I've re-downloaded and re-installed, the test is running right now.
Else I'll try to update the GCC, and otherwise will see if one of the earlier kover versions maybe runs.

Regards,
Bastian

**Comment by bastian-wur** [comment]
I reinstalled v1.3 (which you linked somewhere), now it seems to have worked (did not yet investigate the result though). No clue what helped o_O. But thanks for the support :).

**Comment by aldro61** [comment]
Looks like some kind of compilation issue. Glad that it's fixed. Let me know if you run into any other problems.

Cheers,
Alex

**Comment by bastian-wur** [comment]
FYI: It does not seem to be a compilation issue.
It worked when I first ran it, but now I'm trying again (for other projects, and also rerunning the original one with more data for one of the groups), and I'm getting the same error again (and have not figured out yet how to get around).
It's probably somewhere a problem in the dependencies, but I really have no idea.
I'll fiddle around with it a bit more, maybe I'll find something.

**Comment by aldro61** [comment]
Thanks for getting back to me! Do you have another system (e.g., your personal computer) on which you could try running kover on the same data? If it works, it would confirm that the error is not due to anything data-specific.

**Comment by aldro61** [comment]
I've been thinking about this and I doubt that the problem is data related, since you also failed to run the example. I'll take a look at the code and get back to you. Can you confirm that it is still failing when you call kover dataset split?

**Comment by bastian-wur** [comment]
Hi Alex,

yes, it's still failing at the splitting operation.
I am now installing it on my normal machine at work (not on the cluster), with a fresh installation, and I'll see if the test data works.

Thanks for the help :).

**Comment by bastian-wur** [comment]
Okay, took me some time, but now... so... nothing wrong with kover, the algorithm, or anything itself, so sorry for bothering you, and thanks for the support.

In case anyone wonders: It turns out that the error message "Illegal instruction (core dumped)" means that there was an illegal instruction used on the CPU. Means that the binary was compiled with instructions, which the executing CPU does not understand (these instructions are the flags under /proc/cpuinfo ).
In my case, that was because I work on a cluster, which has all different kinds of CPUs. I compiled on our headnode, which as a CPU, which is a lot newer than any of the CPUs in the cluster. So when my job was executed on any of the nodes on the cluster, it ran on older CPUs, which did not have all the necessary flags, just because they are older.
I fixed the problem (with the help of a few people on my slack channel; there also the initial hint came from), by checking which nodes in he cluster have older CPUs and have the least flags, and compiled on one of them.
Now my job has run, and need to check the results.

Thanks again, and issue closed :).
